server/src/MessageHandler.ts

```diff
@@ -5,77 +5,190 @@ import {
+  return document.getText({
+    start: { line: line, character: 0 },
+    end: { line: line, character: 9999 },
   })
+}
+
+function convertDocumentTextToTrimmedLineArray(
+  document: TextDocument,
+): Array<string> {
+  return [...Array(document.lineCount)].map((_, i) =>
+    getCurrentLine(document, i).trim(),
+  )
+}
+
+function isFirstInsideBlock(position: Position, currentLine: string): boolean {
+  if (currentLine.trim().length === 0) {
+    return true
+  }
+
+  const stringTillPosition = currentLine.slice(0, position.character)
+  const matchArray = new RegExp(/\w+/).exec(stringTillPosition)
+
+  if (!matchArray) {
+    return true
+  }
+  return (
+    matchArray.length === 1 &&
+    matchArray.index !== undefined &&
+    stringTillPosition.length - matchArray.index - matchArray[0].length === 0
+  )
+}
+
+function getWordAtPosition(document: TextDocument, position: Position): string {
+  const currentLine = getCurrentLine(document, position.line)
+  const stringTillPosition = currentLine.slice(0, position.character)
+  if (stringTillPosition.endsWith('@@')) {
+    return '@@'
+  }
+  if (stringTillPosition.endsWith('@')) {
+    return '@'
+  }
   // search for the word's beginning and end
-  const beginning = currentLine.slice(0, position.character + 1).search(/\S+$/)
-  const end = currentLine.slice(position.character).search(/\W/)
+  const beginning: number = currentLine
+    .slice(0, position.character + 1)
+    .search(/\S+$/)
+  const end: number = currentLine.slice(position.character).search(/\W/)
   if (end < 0) {
     return ''
   }
   return currentLine.slice(beginning, end + position.character)
 }
 
-export async function handleDefinitionRequest(
+export class MyBlock {
+  type: string
+  start: Position
+  end: Position
+  name: string
+
+  constructor(type: string, start: Position, end: Position, name: string) {
+    this.type = type
+    this.start = start
+    this.end = end
+    this.name = name
+  }
+}
+
+function getBlockAtPosition(
+  line: number,
+  lines: Array<string>,
+): MyBlock | void {
+  let blockType = ''
+  let blockName = ''
+  let blockStart: Position = Position.create(0, 0)
+  let blockEnd: Position = Position.create(0, 0)
+  // get block beginning
+  let reachedLine = false
+  for (const [key, item] of lines.reverse().entries()) {
+    const actualIndex = lines.length - 1 - key
+    if (actualIndex === line) {
+      reachedLine = true
+    }
+    if (!reachedLine) {
+      continue
+    }
+    if (item.includes('{')) {
+      const index = item.search(/\s+/)
+      blockType = ~index ? item.slice(0, index) : item
+      blockName = item.slice(blockType.length, item.length - 2).trim()
+      blockStart = Position.create(actualIndex, 0)
+      break
+    }
+    // not inside a block
+    if (item.includes('}')) {
+      return
+    }
+  }
+  reachedLine = false
+  // get block ending
+  for (const [key, item] of lines.reverse().entries()) {
```

Comment:
You might want to save the `lines.reverse().entries()` result in a variable and use it in both cases, as both `reverse` and `entries` are expensive operations.

Reply:
I'll probably need to make a deep copy here as well.. I only really need the reversed array in the first case, but since it changes `lines`, I added the second `reverse` statement

Reply:
Ok, then the `reverse` makes sense